feat: Properly bubble up `InsertBlockFatalError`

martinezjorge opened the pull request:
Closes #10219

martinezjorge commented on crates/engine/tree/src/engine.rs:
```diff
@@ -205,6 +206,9 @@ where
             }
             EngineApiEvent::Download(action) => RequestHandlerEvent::Download(action),
+            EngineApiEvent::FatalBlockInsert(error) => {
+                RequestHandlerEvent::HandlerEvent(HandlerEvent::FatalError)
```
I'm not sure if this is the right approach here because other places use HandlerEvent::FatalError but they do not pass an error and I believe the context is slightly different so I'm not sure if I should add another handler event or if the existing one should be modified.

martinezjorge commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -1398,6 +1398,7 @@ where
                     debug!(target: "engine", ?err, "failed to connect buffered block to tree");
                     if let Err(fatal) = self.on_insert_block_error(err) {
                         warn!(target: "engine", %fatal, "fatal error occurred while connecting buffered blocks");
+                        self.emit_event(EngineApiEvent::FatalBlockInsert(fatal));
```
Not sure if this is how the fatal error is meant to be handled.

martinezjorge commented:
@Rjected Could I get some feedback/direction on this approach?

Rjected reviewed (changes requested):
instead of using `emit_event`, let's ensure that the methods which call `on_insert_block_error` return a `Result` so we can do
```rust
self.on_insert_block_error(err)?;
```

does that make sense?

martinezjorge commented:
@Rjected I switched from bubbling towards main thread using emits to bubbling up in `EngineApiTreeHandler` until the `fn run`. I also updated the tests. Please let me know you'd like me to approach the bubble up or the tests differently or if there is a test you'd like me to add.

Rjected commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -885,21 +887,30 @@ where
                 }
                 FromOrchestrator::BackfillSyncFinished(ctrl) => {
-                    self.on_backfill_sync_finished(ctrl);
+                    if let Err(fatal) = self.on_backfill_sync_finished(ctrl) {
+                        Err(fatal)
+                    } else {
+                        Ok(())
+                    }
```
```suggestion
                    self.on_backfill_sync_finished(ctrl)?;
```

Rjected commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -937,22 +949,27 @@ where
             }
-            FromEngine::DownloadedBlocks(blocks) => {
-                if let Some(event) = self.on_downloaded(blocks) {
+            FromEngine::DownloadedBlocks(blocks) => match self.on_downloaded(blocks) {
```
this can be
```rust
FromEngine::DownloadedBlocks(blocks) => {
    if let Some(event) = self.on_downloaded(blocks)? {
        self.on_tree_event(event);
    }
}
```

Rjected commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -526,22 +525,25 @@ where
         let batch = self.config.max_execute_block_batch_size().min(blocks.len());
         for block in blocks.drain(..batch) {
-            if let Some(event) = self.on_downloaded_block(block) {
+            if let Ok(Some(event)) = self.on_downloaded_block(block) {
```
```suggestion
            if let Some(event) = self.on_downloaded_block(block)? {
```

martinezjorge replied:
The issue I'm having with this is that `on_downloaded_block` can return an InsertBlockFatalError but it doesn't return an option on success and as such the compiler won't let me use `?` operator inside of `run`.

Rjected commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -500,24 +500,23 @@ where
     ///
     /// This will block the current thread and process incoming messages.
-    pub fn run(mut self) {
+    pub fn run(mut self) -> Result<(), InsertBlockFatalError> {
```
in this method I would prefer that the signature remain as `run(mut self)`, and instead we handle the `on_engine_message` like this:
```rust
if let Err(err) = self.on_engine_message(msg) {
    // some error! message
    return
}
```

martinezjorge replied:
Okay no problem, I was not sure about this detail. Thank you for clarifying.

Rjected commented on crates/engine/tree/src/chain.rs:
```diff
@@ -189,7 +189,7 @@ pub trait ChainHandler: Send + Sync {
 
 /// Events/Requests that the [`ChainHandler`] can emit to the [`ChainOrchestrator`].
-#[derive(Clone, Debug)]
+#[derive(Debug)]
```
we can still have `Clone` here right?

martinezjorge replied:
Yeah absolutely, that was from the previous change but I was curious what you would say about it.

Rjected reviewed (changes requested):
nice! some comments on result handling and the `run` signature

martinezjorge commented:
Made the requested changes but I'm having a couple unit tests that are hanging

> test tree::tests::test_engine_tree_live_sync_transition_eventually_canonical has been running for over 60 seconds
> test tree::tests::test_engine_tree_live_sync_transition_required_blocks_requested has been running for over 60 seconds

So investigating, trying to figure out what's going on with those

martinezjorge commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -518,7 +518,10 @@ where
-                    self.on_engine_message(msg);
+                    if let Err(fatal) = self.on_engine_message(msg) {
+                        error!(target: "engine", %fatal, "insert block fatal error");
+                        // return
```
If it breaks it'll just loop again

and

If it returns we bail out of run()?

I'm leaning towards break but I'm not sure which behavior we want on `InsertBlockFatalError`

Rjected replied:
> If it breaks it'll just loop again

only `continue` will continue to loop again, `break` here should be the same as `return`, since there is only one level of loop

martinezjorge replied:
ah okay thanks for clarifying

Rjected commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -169,7 +169,7 @@ impl TreeState {
         while let Some(current_block) = self.block_by_hash(current_hash) {
             if current_block.hash() == target_hash {
-                return false
+                return false;
```
one nit: let's remove the added semicolons

martinezjorge replied:
no problem

Rjected commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -518,20 +518,23 @@ where
+                        error!(target: "engine", %fatal, "insert block fatal error");
+                        break;
+                    }
                 }
                 Ok(None) => {
                     debug!(target: "engine", "received no engine message for some time, while waiting for persistence task to complete");
                 }
                 Err(_err) => {
                     error!(target: "engine", "Engine channel disconnected");
-                    return
+                    return;
                 }
             }
 
             if let Err(err) = self.advance_persistence() {
                 error!(target: "engine", %err, "Advancing persistence failed");
-                break
+                break;
```
let's `return` here, they should be equivalent but this is just to stay consistent

martinezjorge replied:
gotcha

Rjected commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -3000,16 +3046,18 @@ mod tests {
         test_harness.fcu_to(main_chain_last_hash, ForkchoiceStatus::Syncing).await;
 
         let event = test_harness.from_tree_rx.recv().await.unwrap();
+        println!("event received: {:?}", event);
```
```suggestion
```

martinezjorge replied:
whoops

Rjected commented on crates/engine/tree/src/tree/mod.rs:
```diff
@@ -2919,12 +2958,16 @@ mod tests {
+            .unwrap();
 
         let event = test_harness.from_tree_rx.recv().await.unwrap();
+        println!("event received: {:?}", event);
```
```suggestion
```

martinezjorge replied:
whoops x 2

Rjected reviewed (changes requested):
Nice! I have some small nitpicks but this seems good otherwise

Rjected reviewed (approved):
This looks good to me, thanks!
